### ISSUE TITLE
Quantile or Jenks classification

### ISSUE THREAD

**Issue by sylvaincaillot** [humangeo/leaflet-dvf#55]
Good morning, 
First thanks for your library which produces amazing maps.
I have an issue with some recent data for which the current classification does not render it properly in one of my chloropeth map (few records with extreme values).
I think it is related to the fact that the current classification I am using with DVF is of equal interval or defined interval.

Would it be possible to choose a different classification such as Jenks or quantile to properly render my data and maximize the variance between my classes?

Thanks in advance for your answer

Sylvain

**Comment by sfairgrieve** [comment]
That sounds like a great idea.  I'll give it some thought.  If you have any suggestions on how to best implement this, let me know.  I've been checking out [TurfJS](https://github.com/Turfjs/turf) recently, which seems like a good fit here possibly.

**Comment by sylvaincaillot** [comment]
Thanks Scott, 
I have used this piece of javascript code: https://gist.github.com/tmcw/4977508 from the creator of turf to create jenks classes. It works great.
I have now the boundaries of my classes, 
For examples, i have 3 classes with the following breaks :[1712, 2021, 8855, 20421]

How can i inject these values into the display options of my map to have relevant color palette and legend?

Thanks for your answer.

Sylvain

**Comment by sfairgrieve** [comment]
Sylvain,

Currently, you can use an _L.PiecewiseFunction_ instance to specify the color function.  To use _L.PiecewiseFunction_, you pass in an array of _LinearFunction_ instances (e.g. color functions), like so:  

```javascript
var colorFunction = new L.PiecewiseFunction([
   new L.HSLHueFunction([1712,  0], [2021, 30]),
   new L.HSLHueFunction([2021, 30], [8855, 60]),
   new L.HSLHueFunction([8855, 60], [20421, 90])
]);
```

This approach gives you complete control over color output, but can be a little involved for this general use case.  I need to make some tweaks to the library to make this easier.  I'm thinking of at least adding a class that takes an array of breaks and an array of colors and does this for you behind the scenes.  

In terms of the legend, I also need to make some updates.  Currently, the legend will display some fixed number of boxes (controllable via the _legendOptions.numSegments_ option), where each box has a fixed width (_legendOptions.width_ / _legendOptions.numSegments_).  The default width is 100px and the default numSegments is 10, so each box is 10px wide.  The legend will work for this use case, but it's not optimal.  Based on other legends I've seen, my feeling is that it needs to be updated to only a display a box for each class.  For now, you can get closer to this effect by removing the border-style: solid CSS style from each legend element - this should at least give the legend more of a gradient appearance where it isn't so obvious that there's not a box for each class.

I'll let you know when I get in enhancements to address this, but I'm hoping in the meantime the current approach will work for your use case.  Let me know what you think and if I can help further.

**Comment by sylvaincaillot** [comment]
Thanks a lot Scott,
This is what I was thinking, i tried to create a "L.CustomJenksColorFunction" which was an extension of PiecewiseFunction and which was taking breaks and color as parameters but with no success. I will be using your code in the meantime and then switch to the next version when ready.

Thanks again for all

Sylvain

**Comment by sfairgrieve** [comment]
Sylvain,

I just pushed a branch called classification that has a new function called L.ColorClassFunction (I'm open to changing the name).  You should be able to pass this an array of breaks along with an array of colors (per your example).  Let me know if this works for you, and I'll merge it into master.  Here's the code for the new class:

```javascript
L.ColorClassFunction = L.PiecewiseFunction.extend({
	options: {
		interpolate: true
	},
	
	initialize: function (classBreaks, colors, options) {
		var functions = [];
		var colorFunction;
		
		L.Util.setOptions(this, options);
		
		for (var i = 0; i < classBreaks.length - 1; ++i) {
			var start = classBreaks[i],
				end = classBreaks[i + 1],
				startColor = colors[i],
				endColor = this.options.interpolate ? colors[Math.min(colors.length -1, i + 1)] : colors[i];
			
			colorFunction = new L.RGBColorBlendFunction(start, end, startColor, endColor);
			
			functions.push(colorFunction);	
		}
		
		L.PiecewiseFunction.prototype.initialize.call(this, functions);
	}
});
```

Scott

**Comment by sylvaincaillot** [comment]
Thank you so much Scott,

I just tried it and it looks promising.
'''
var breaks = getJenks(dataJSON.grid[0], 'A', 4);	          

            var fillColor = new L.ColorClassFunction(breaks, L.ColorBrewer.Diverging[selectedPalette][4].slice(0).reverse(), {
                interpolate: false
            });
'''
I just see an issue of color between the legend and what is shown on the map:
Please see the difference below between normal L.CustomColorFunction (equal intervals) and the new L.ColorClassFunction (Jenks) 
![dvf_jenks](https://cloud.githubusercontent.com/assets/4845910/6322057/b44fd916-bb60-11e4-9bd4-52f1c4408b73.png)

In my case, i am dealing with 4 classes, the Jenks classes are [166, 789, 1161, 1685, 2540] but when I am mousing over the legend to display boundaries of each color class, it shows values corresponding to equal value classification ([166, 759.5, 1353, 1946.5, 2540] with an interval of 593.5).
Moreover, the green color shown in the map is not displayed in the legend and the red color is over 2 classes which is a bit confusing.

Hope it make sense and thanks again

Sylvain

**Comment by sfairgrieve** [comment]
That makes sense, and I definitely see the issue.  I'll make some updates to the legend building code that will hopefully solve this issue.

**Comment by sfairgrieve** [comment]
I think I have a fix for the legend, but I just need to do some more testing/code cleanup before I push it.  I'll push it to the classification branch at some point today once I get some time to work on it.

**Comment by sfairgrieve** [comment]
Went ahead and pushed it to the classification branch.  I didn't do a ton of testing, so let me know if you encounter any issues.  In order for the legend to work with breaks, you'll need to supply the breaks as part of the legendOptions you pass in to the DataLayer like so:

```javascript
var dataLayer = new L.ChoroplethDataLayer(data, {
    ...
    legendOptions: {
        breaks: <array of break values>
    }
});
```

**Comment by sfairgrieve** [comment]
Just realized that this solution won't work for all cases but should work for your use case.  Need to make some more tweaks before it's ready for release.

**Comment by sfairgrieve** [comment]
Ok one more update, now pass breaks with your displayOptions field like so:

```javascript
var dataLayer = new L.ChoroplethDataLayer(data, {
    ...
    displayOptions: {
        'field1': {
            ...
            breaks: <array of break values>
        }
    }
});
```

**Comment by sylvaincaillot** [comment]
This is great Scott.
This is my current code using your last update:
```
var breaks = getJenks(dataJSON.grid[0], MyField, parseInt(MyClass, 10));	     

            var fillColor = new L.ColorClassFunction(breaks, L.ColorBrewer.Diverging[MyPalette][MyClass].slice(0).reverse(), {
                interpolate: false
            });

            var color = new L.ColorClassFunction(breaks, L.ColorBrewer.Diverging[MyPalette][MyClass].slice(0).reverse(), {
                postProcess: function (y) {
                    var newColor = new L.RGBColor(y);
                    newColor.l(0.2);
                    return newColor.toRGBString();
                }
            });

            displayOptions[MyField] = {
                displayName: $('#field-select option:selected').text(),
                fillColor: fillColor,
                color: color,
                breaks: breaks
            }

MyLeafletLayer.setDisplayOptions(displayOptions);
```
I think we are almost there, the legend is correctly generated in term of boundaries values and class width. I still have this issue of color being duplicated. As you can see below, the green color has disappeared from the legend and the red has been duplicated which makes the legend wrong as the green color still exists in the map.
I tried to explore the _getLegendElement element  in your code where the color is allocated to each legend class with no success
![dvf_jenks_2](https://cloud.githubusercontent.com/assets/4845910/6386607/bba8c224-bdcf-11e4-84af-5b71cd720b6f.png)

Thanks for everything

Sylvain

**Comment by sfairgrieve** [comment]
Thanks Sylvain.  Looking into this now.  What are the break values for your example above?

**Comment by sylvaincaillot** [comment]
Sure. My break values with 5 classes are [166, 729, 929, 1241, 1713, 2540]

Thanks again
S.

**Comment by sfairgrieve** [comment]
Sylvain, give the latest a try when you get a chance.  I believe it should work, but appreciate your testing it.  Thanks for your patience.  I definitely need to spend some time rethinking/refactoring the legend code.

**Comment by sylvaincaillot** [comment]
Thanks Scott,
Everything works perfectly for my break classification and the legend display now.
The last issue to be corrected, which is still there in the master version, is the fact that the chloropeth map colors do not match the default legend.

Let me explained: In the case of a default legend (equal intervals) with 5 classes for example (boundaries: 166, 640, 1115, 1590, 2065, 2540 with colors: blue, green, yellow, orange, red), a square with the value 2300 will be shown in orange while it should be red.
Please see below an example for which the squares of the grid appears in yellow while they should be in orange since their values is between 1590.4 and 2065.2:
![dvf_equal](https://cloud.githubusercontent.com/assets/4845910/6435013/885bcbac-c0e8-11e4-8845-bce3b9a76e66.png)

When the classification chosen is Jenks with breaks, the legend and colored grid matched perfectly.

Hope it makes sense and thanks again for the push concerning the breaks classification.

Sylvain

**Comment by sfairgrieve** [comment]
Almost there.  Thanks for sticking with this.  I'll look into it and hopefully have a fix in soon.

**Comment by sfairgrieve** [comment]
Does the latest commit fix this?  I noticed/fixed an issue with the legend code.  If not, does it work when you pass in the equal interval breaks yourself (you might already be doing this?):

```javascript
var breaks = [166, 640, 1115, 1590, 2065, 2540];
var colors = L.ColorBrewer.Diverging['Spectral'][5].slice(0).reverse();
var colorFunction = new L.ColorClassFunction(breaks, colors);
var layer = new L.DataLayer(data, {
     ...
     displayOptions: {
         <field>: {
              fillColor: colorFunction,
              breaks: breaks
         }
     }
});
```

**Comment by sylvaincaillot** [comment]
Thanks Scott,
The breaks classification works fine since your previous release and yes, I am already passing the interval breaks myself as you code is showing it.
It is only the default classification which is not working properly. Now the colors are spread over the legend with 2 classes with the same color and the red color missing (see below). It also seems that there is an inconsistency between the color given in the map and the range values indicated in the legend.
![equal_interval_error_2](https://cloud.githubusercontent.com/assets/4845910/6472738/a39304d8-c248-11e4-8e98-59532bd8d1bb.png)

Sorry about that

Sylvain

**Comment by sfairgrieve** [comment]
Darn.  Would you be willing to share your code (or an equivalent example when classification is set to equal intervals) and some example data?  I understand if you don't want to.  Just want to try and recreate this issue and make sure my test case matches your case.   I've been testing updates against the earthquake example (updated to use the ColorClassFunction and breaks) and it seems to be working fine in the default and jenks case.

**Comment by sfairgrieve** [comment]
I'll add a new example to the classification branch that matches what I think your use case is, and maybe you can let me know if it's correct or not.  Are you passing in legendOptions.numSegments?  I see an issue when I pass in legendOptions.numSegments and don't include the breaks under a field in displayOptions.  The legend is definitely off in that case.

**Comment by sylvaincaillot** [comment]
Thnaks Scott for all of this, 
 i am effectively using the numSegments to define th number of classes.
Please see below part of my code:

This part is generating the default map based on equal intervals classification
````
    selectedClass = 9;
 
    selectedBiomass = 'a';
    selectedPalette = 'Spectral';
    selectedYear = 0;
    selectedType = 'E';
    var range = $.map(dataJSON.grid[0]["_" + selectedBiomass].split('-'), Number);   

    //Diverging
	var fillColor = new L.CustomColorFunction(range[0], range[1], L.ColorBrewer.Diverging[selectedPalette][selectedClass].slice(0).reverse(), {
	    interpolate: false
	});  

    //Diverging
	var color = new L.CustomColorFunction(range[0], range[1], L.ColorBrewer.Diverging[selectedPalette][selectedClass].slice(0).reverse(), {
	    postProcess: function (y) {
	        var newColor = new L.RGBColor(y);
	        newColor.l(0.2);
	        return newColor.toRGBString();
	    }
	});

	var options = {
	    locationMode: L.LocationModes.LOOKUP,
	    recordsField: null,
	    codeField: 'id',		//field in the data Json identifying uniquely a feature to be linked with an item in the GeoJSON
	    locationLookup: gridJSON, // A GeoJSON FeatureCollection that will be used to lookup boundaries/location
	    locationTextField: 'xy', //field in the data Json to be displayed in the mouse hover popup
	    locationIndexField: 'ID', //Property in GeoJSON to identify each feature
	    includeBoundary: true, // Whether or not to include a background boundary so people know what boundary each marker is associated with
	    layerOptions: {
	        fillOpacity: 0.9,
	        opacity: 0.5,
	        weight: 0.5
	    },
	    tooltipOptions: {
	        iconSize: new L.Point(130, 60),
	        iconAnchor: new L.Point(-5, 60)
	    },
	    legendOptions: {},
	    displayOptions: {}
	};

	options.displayOptions[selectedBiomass] = {
	    displayName: $('#biomass-select option:selected').text(),
	    fillColor: fillColor,
	    color: color
	};

	options.legendOptions[selectedBiomass] = {
	    numSegments: selectedClass,
	    gradient: false
	};

	biomassLayer = new L.ChoroplethDataLayer(dataJSON.grid[0].items, options);

	layerControl.addOverlay(biomassLayer, 'Biomass');
	
	map.addLayer(biomassLayer);
````

Then the user has the possibility to change the palette, the classification type (Jenks or equals interval, number of classes), see below the generic function which handle that with the function when number of classes are changed :
'''
    $classSelect.on('change', function (e) {

        selectedClass = $classSelect.val();
        var displayOptions = updateColor();

        //https://github.com/humangeo/leaflet-dvf/issues/26
        biomassLayer.options.legendOptions.numSegments = selectedClass;
        legendControl.redrawLayer(biomassLayer);

        biomassLayer.setDisplayOptions(displayOptions);

    });
function updateColor() {
        var displayOptions = {};

        if (selectedType == 'E') {
            //--------------------------------------------   Equal intervals
            var range = $.map(dataJSON.grid[selectedYear]["_" + selectedBiomass].split('-'), Number);

            var fillColor = new L.CustomColorFunction(range[0], range[1], L.ColorBrewer.Diverging[selectedPalette][selectedClass].slice(0).reverse(), {
                interpolate: false
            });

            var color = new L.CustomColorFunction(range[0], range[1], L.ColorBrewer.Diverging[selectedPalette][selectedClass].slice(0).reverse(), {
                postProcess: function (y) {
                    var newColor = new L.RGBColor(y);
                    newColor.l(0.2);
                    return newColor.toRGBString();
                }
            });

            displayOptions[selectedBiomass] = {
                displayName: $('#biomass-select option:selected').text(),
                fillColor: fillColor,
                color: color
            }

        } else {
            //--------------------------------------------   Natural breaks (Jenks)
            var breaks = getJenks(dataJSON.grid[selectedYear], selectedBiomass, parseInt(selectedClass, 10));	
            alert(JSON.stringify(breaks));

            var fillColor = new L.ColorClassFunction(breaks, L.ColorBrewer.Diverging[selectedPalette][selectedClass].slice(0).reverse(), {
                interpolate: false
            });

            var color = new L.ColorClassFunction(breaks, L.ColorBrewer.Diverging[selectedPalette][selectedClass].slice(0).reverse(), {
                postProcess: function (y) {
                    var newColor = new L.RGBColor(y);
                    newColor.l(0.2);
                    return newColor.toRGBString();
                }
            });

            displayOptions[selectedBiomass] = {
                displayName: $('#biomass-select option:selected').text(),
                fillColor: fillColor,
                color: color,
                breaks: breaks
            }
        } 
        return displayOptions;
    }
'''
Again everything seems fine for the breaks classification, it is really the default classsifcation which is giving me a hard time.
Hope it is detailed enough. If you give me your contact details, i could even give you  the credentials to access my application and see the issue in live.

Thanks again

Sylvain

**Comment by sfairgrieve** [comment]
Thanks for sharing.  I think I understand what the problem is.  There's an issue with the CustomColorFunction.  Should have a fix in soon.  Hopefully this actually fixes things.

**Comment by sylvaincaillot** [comment]
Scott, 
I just tried the last version of the classification branch you pushed lately and it works beautifully.

Thank you very much.
I have just added a rounding to the min and max values of the legend boundaries in the _getLegendElement function as I had numbers with extreme decimal (125.200000000003) as shown in one of my previous post.

It is perfect now.
Thanks again for your precious help and your reactivity

Sylvain
